scanner/services/indexer.go

```diff
@@ -0,0 +1,140 @@
+}
+
+// NewIndexerService creates a new indexer service.
+func NewIndexerService(indexer indexer.Indexer) *indexerService {
+	return &indexerService{
+		indexer: indexer,
+	}
+}
+
+func (s *indexerService) CreateIndexReport(ctx context.Context, req *v4.CreateIndexReportRequest) (*v4.IndexReport, error) {
+	// TODO We currently only support container images, hence we assume the resource
+	//      is of that type. When introducing nodes and other resources, this should
+	//      evolve.
+	resourceType := "containerimage"
+	if err := validateContainerImageRequest(req); err != nil {
+		return nil, err
+	}
+	// Create a claircore.Digest using the full Hash ID, including resource type.
+	hashIDSum := sha512.Sum512([]byte(req.GetHashId()))
+	manifestDigest, err := claircore.NewDigest(claircore.SHA512, hashIDSum[:])
+	if err != nil {
+		return nil, fmt.Errorf("internal error: creating container image manifest digest: %w", err)
+	}
+
+	ctx = zlog.ContextWithValues(ctx,
+		"component", "scanner/service/indexer",
+		"resource_type", resourceType,
+		"manifest_digest", manifestDigest.String())
+
+	// Setup authentication.
+	var opts []indexer.Option
+	hasAuth := req.GetContainerImage().GetUsername() != ""
+	if hasAuth {
+		opts = append(opts, indexer.WithAuth(&authn.Basic{
+			Username: req.GetContainerImage().GetUsername(),
+			Password: req.GetContainerImage().GetPassword(),
+		}))
+	}
+
+	// Create index report.
+	zlog.Info(ctx).
+		Str("image_url", req.GetContainerImage().GetUrl()).
+		Bool("has_auth", hasAuth).
+		Msg("creating index report for container image")
+	clairReport, err := s.indexer.IndexContainerImage(
+		ctx,
+		manifestDigest,
+		req.GetContainerImage().GetUrl(),
+		opts...)
+	if err != nil {
+		zlog.Error(ctx).Err(err).Send()
+		return nil, err
+	}
+	indexReport := convertToIndexReport(clairReport)
+	indexReport.HashId = req.GetHashId()
+	// TODO Define behavior for indexReport.Err != "".
+	return indexReport, nil
+}
+
+// validateContainerImageRequest validates a container image request.
+func validateContainerImageRequest(req *v4.CreateIndexReportRequest) error {
+	if req == nil {
+		return errox.InvalidArgs.New("empty request")
+	}
+	if !strings.HasPrefix(req.GetHashId(), "/v4/containerimage/") {
+		return errox.InvalidArgs.Newf("invalid hash id: %q", req.GetHashId())
+	}
+	if req.GetContainerImage() == nil {
+		return errox.InvalidArgs.New("invalid resource locator for container image")
+	}
+	// Validate container image URL.
+	imgURL := req.GetContainerImage().GetUrl()
+	if imgURL == "" {
+		return errox.InvalidArgs.New("missing image URL")
+	}
+	u, err := url.Parse(imgURL)
+	if err != nil {
+		return errox.InvalidArgs.Newf("invalid image URL: %q", imgURL).CausedBy(err)
+	}
+	if u.Scheme != "http" && u.Scheme != "https" {
+		return errox.InvalidArgs.New("image URL does not start with http:// or https://")
+	}
+	imageRef := strings.TrimPrefix(imgURL, u.Scheme+"://")
+	_, err = name.ParseReference(imageRef, name.StrictValidation)
```

Comment:
elsewhere we added the insecure option. Is it not needed here?

Reply:
I don't know. Can you point out where and why it could be needed here?

Reply:
It is used in scanner/indexer/indexer.go on line 213. However, it is just useful when actually making requests with the returned reference. Here, we just create the reference for validation purposes. Can you maybe add a comment saying why we don't need to consider adding the insecure option here?

Reply:
I see. And you understand it correctly. We only use it for validation here. Whether secure or insecure is irrelevant. It doesn't warrant a comment, though. Isn't the context enough?